File: 8.string-to-integer-atoi/solution.py
class Solution:
    def myAtoi(self, str: str) -> int:
        str = str.lstrip()
        i = 0
        sig = 1
        dict = {"0": 0, "1": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9}
        if len(str) == 0:
            return 0
        if str[0] == "+":
            i = 1
        elif str[0] == "-":
            i = 1
            sig = -1
        result = 0
        while i < len(str):
            if str[i] not in dict:
                break
            result = result * 10 + dict[str[i]] * sig
            if result >= 2147483648:
                return 2147483647
            if result < -2147483648:
                return -2147483648
            i += 1
        return result

File: 8.string-to-integer-atoi/solution.go
func myAtoi(s string) int {
	s = strings.Trim(s, " ")
	for i := 1; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			s = s[0:i]
			break
		}
	}
	if len(s) == 0 {
		return 0
	}
	var sig int64 = 1
	if s[0] == '-' {
		sig = -1
		s = s[1:]
	} else if s[0] == '+' {
		s = s[1:]
	}
	for i := 0; i < len(s); i++ {
		if s[i] != '0' {
			s = s[i:]
			break
		} else if i == len(s)-1 {
			return 0
		}
	}
	if len(s) > 12 {
		if sig == -1 {
			return math.MinInt32
		}
		return math.MaxInt32
	}
	val, err := strconv.ParseInt(strings.Trim(s, " "), 10, 64)
	if err != nil {
		return 0
	}
	val *= sig
	if val > math.MaxInt32 {
		return math.MaxInt32
	} else if val < math.MinInt32 {
		return math.MinInt32
	}
	return int(val)
}
